From 5ce44e4071ce35616a7a6b2125c43acc73f18fca Mon Sep 17 00:00:00 2001
From: Andreas Schultz <andreas.schultz@travelping.com>
Date: Thu, 18 Jul 2024 09:33:57 +0200
Subject: [PATCH 35/35] TCP/UDP pseudo header checksum must be in network order

Converting them from network to host order is wrong. The kernel
expects a valid, partial TCP/UDP header checksum in network byte order
and not in host by order.

The method is used only by the af_packet and virtio drivers.
---
 src/vnet/ip/ip_psh_cksum.h | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/vnet/ip/ip_psh_cksum.h b/src/vnet/ip/ip_psh_cksum.h
index 8723749..6de0286 100644
--- a/src/vnet/ip/ip_psh_cksum.h
+++ b/src/vnet/ip/ip_psh_cksum.h
@@ -38,8 +38,7 @@ ip4_pseudo_header_cksum (ip4_header_t *ip4)
   psh.proto = ip4->protocol;
   psh.l4len = clib_host_to_net_u16 (clib_net_to_host_u16 (ip4->length) -
 				    sizeof (ip4_header_t));
-  return ~clib_net_to_host_u16 (
-    clib_ip_csum ((u8 *) &psh, sizeof (ip4_psh_t)));
+  return ~clib_ip_csum ((u8 *) &psh, sizeof (ip4_psh_t));
 }
 
 static_always_inline u16
@@ -50,8 +49,7 @@ ip6_pseudo_header_cksum (ip6_header_t *ip6)
   psh.dst = ip6->dst_address;
   psh.l4len = ip6->payload_length;
   psh.proto = clib_host_to_net_u32 ((u32) ip6->protocol);
-  return ~clib_net_to_host_u16 (
-    clib_ip_csum ((u8 *) &psh, sizeof (ip6_psh_t)));
+  return ~clib_ip_csum ((u8 *) &psh, sizeof (ip6_psh_t));
 }
 
 #endif /* included_ip_psh_cksum_h */
-- 
2.43.0

